core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
+                           txnMarkerResult: TransactionResult,
                            responseCallback: EndTxnCallback): Unit = {
-    val errors = validateTransactionalId(transactionalId)
-    if (errors != Errors.NONE)
-      responseCallback(errors)
-    else
-      txnManager.getTransactionState(transactionalId) match {
+    val error = validateTransactionalId(transactionalId)
+    if (error != Errors.NONE)
+      responseCallback(error)
+    else {
+      val result: Either[Errors, (Int, TransactionMetadataTransition)] = txnManager.getTransactionState(transactionalId) match {
         case None =>
-          responseCallback(Errors.INVALID_PID_MAPPING)
-        case Some(metadata) =>
-          metadata synchronized {
-            if (metadata.pid != pid)
-              responseCallback(Errors.INVALID_PID_MAPPING)
-            else if (metadata.producerEpoch != epoch)
-              responseCallback(Errors.INVALID_PRODUCER_EPOCH)
-            else metadata.state match {
+          Left(Errors.INVALID_PID_MAPPING)
+
+        case Some(epochAndTxnMetadata) =>
+          val txnMetadata = epochAndTxnMetadata.transactionMetadata
+          val coordinatorEpoch = epochAndTxnMetadata.coordinatorEpoch
+
+          txnMetadata synchronized {
+            if (txnMetadata.producerId != pid)
+              Left(Errors.INVALID_PID_MAPPING)
+            else if (txnMetadata.producerEpoch != epoch)
+              Left(Errors.INVALID_PRODUCER_EPOCH)
+            else if (txnMetadata.pendingTransitionInProgress)
+              Left(Errors.CONCURRENT_TRANSACTIONS)
+            else txnMetadata.state match {
               case Ongoing =>
-                commitOrAbort(transactionalId, pid, epoch, command, responseCallback, metadata)
+                val nextState = if (txnMarkerResult == TransactionResult.COMMIT)
+                  PrepareCommit
+                else
+                  PrepareAbort
+                Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, time.milliseconds()))
               case CompleteCommit =>
-                if (command == TransactionResult.COMMIT)
-                  responseCallback(Errors.NONE)
+                if (txnMarkerResult == TransactionResult.COMMIT)
+                  // this is an optimization: return OK immediately
+                  Left(Errors.NONE)
                 else
-                  responseCallback(Errors.INVALID_TXN_STATE)
+                  Left(Errors.INVALID_TXN_STATE)
               case CompleteAbort =>
-                if (command == TransactionResult.ABORT)
-                  responseCallback(Errors.NONE)
+                if (txnMarkerResult == TransactionResult.ABORT)
+                // this is an optimization: return OK immediately
+                  Left(Errors.NONE)
                 else
-                  responseCallback(Errors.INVALID_TXN_STATE)
+                  Left(Errors.INVALID_TXN_STATE)
               case _ =>
```

Comment:
Hmm... Isn't it possible that a client resends an EndTxnRequest while we are still in `PrepareCommit` or `PrepareAbort`. As long as the outcome matches, it seems we should accept those requests and perhaps return CONCURRENT_TRANSACTIONS?

Also, can we list the states we're catching here explicitly?